File: 프로그래머스/3/43162. 네트워크/네트워크.cpp
#include <string>
#include <vector>

using namespace std;
vector<vector<int>>cc;
int visit[205], arr[205][205];
int cn,answer = 0;

void dfs(int r){
    visit[r] = 1;
    for(int i = 0; i<cn; i++){
        if(cc[r][i] && !visit[i]) dfs(i);
    }
}

int solution(int n, vector<vector<int>> c) {
    cc = c;
    cn = n;
    
    for(int i = 0; i<n; i++){
        for(int j = 0; j<n; j++) if(c[i][j]) arr[i][j] = 1;
    }
    
    for(int i = 0; i<n; i++){
        if(!visit[i]){
            answer++;
            dfs(i);
        }
    }
    return answer;
}

File: 프로그래머스/3/43162. 네트워크/네트워크.java
import java.util.*;

class Solution {
    static List<Integer> adj[];
    static boolean[] visited;
    static int answer;
    public int solution(int n, int[][] computers) {
        init(n, computers);
        for(int i = 0; i < n; i++){
            if(visited[i]) continue;
            answer++;
            dfs(i);
        }
        return answer;
    }
    
    public void dfs(int k){
        visited[k] = true;
        for(int nxt : adj[k]){
            if(!visited[nxt]) dfs(nxt);
        }
    }
    
    public void init(int n, int[][] computers){
        adj = new ArrayList[n];
        visited = new boolean[n];
        for(int i = 0; i < n; i++) adj[i] = new ArrayList<>();
        for(int i = 0; i < n; i++){
            for(int j = 0; j < n; j++){
                if(i == j) continue;
                if(computers[i][j] == 1) adj[i].add(j);
            }
        }
    }
}